src/System.CommandLine.StaticCompletions/shells/ZshShellProvider.cs

```diff
@@ -0,0 +1,371 @@
+
+
+        writer.WriteLine($"#compdef {binaryName}");
+        writer.WriteLine();
+        writer.WriteLine("autoload -U is-at-least");
+        writer.WriteLine();
+        // TODO: if the CLI grammar doesn't support option bundling, remove -s from these options. -s is the bundling option for _arguments
+        writer.WriteLine($$"""
+_{{binaryName}}() {
+    typeset -A opt_args
+    typeset -a _arguments_options
+    local ret=1
+
+    if is-at-least 5.2; then
+        _arguments_options=(-s -S -C)
+    else
+        _arguments_options=(-s -C)
+    fi
+
+    local context curcontext="$curcontext" state state_descr line
+""");
+
+        writer.Indent++;
+        writer.WriteLine(ArgumentsHandler());
+        writer.Indent++;
+        GenerateOptionsAndArgumentsForCommand(pathToCurrentCommand, command, writer);
+        writer.Indent--;
+
+        // tiny hack here - for dynamic completions we need to know what the entire command line is,
+        // so we stash it in a variable and then use it in the dynamic completion handlers
+        writer.WriteLine($$"""local original_args="{{binaryName}} ${line[@]}" """);
+
+        writer.Indent--;
+        writer.Indent++;
+        GenerateSubcommandList(pathToCurrentCommand, command, writer);
+        writer.Indent--;
+        writer.WriteLine("}");
+        writer.WriteLine();
+
+        GenerateSubcommandHandlers(pathToCurrentCommand, command, writer);
+        writer.Indent--;
+        writer.WriteLine($$"""
+if [ "$funcstack[1]" = "_{{binaryName}}" ]; then
+    _{{binaryName}} "$@"
+else
+    compdef _{{binaryName}} {{binaryName}}
+fi
+""");
+        writer.Flush();
+        return textWriter.ToString();
+    }
+
+    private static void GenerateOptionsAndArgumentsForCommand(string[] commandPathForThisCommand, CliCommand command, IndentedTextWriter writer)
+    {
+        var shouldWriteDynamicCompleter = false;
+        foreach (var option in command.HierarchicalOptions())
+        {
+            var multiplicity = option.Arity.MaximumNumberOfValues > 1 ? "*" : "";
+            var helpText = SanitizeHelp(option.Description);
+            if (option.IsFlag())
+            {
+                foreach (var name in option.Names())
+                {
+                    writer.WriteLine($"'{multiplicity}{name}[{helpText}]' \\");
+                }
+            }
+            else
+            {
+                if (option is IDynamicOption)
+                {
+                    shouldWriteDynamicCompleter = true;
+                }
+                var argumentName = option.HelpName ?? " ";
+                var argumentValues = ZshValueExpression(option);
+                foreach (var name in option.Names())
+                {
+                    writer.Write($"'{multiplicity}{name}=[{helpText}]:{argumentName}");
+                    WriteValueExpression(writer, argumentValues);
+                }
+            }
+        }
+
+        var catch_all_emitted = false;
```

Comment:
nit: Is this some kind of contextually named variable? I see others below using this underscore naming style tool.

Reply:
You're right the style is different here - I was transliterating a lot of this from clap-complete, and so I mimicked the code structure and naming of that library to help ease the porting. A subsequent style refactoring could happen, but at least during the writing the similarity helped me keep my place.